applicationset/services/scm_provider/gitea.go

```diff
@@ -0,0 +1,136 @@
+	if insecure {
+		cookieJar, _ := cookiejar.New(nil)
+
+		httpClient = &http.Client{
+			Jar: cookieJar,
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			}}
+	}
+	client, err := gitea.NewClient(url, gitea.SetToken(token), gitea.SetHTTPClient(httpClient))
+	if err != nil {
+		return nil, err
+	}
+	return &GiteaProvider{
+		client:      client,
+		owner:       owner,
+		allBranches: allBranches,
+	}, nil
+}
+
+func (g *GiteaProvider) GetBranches(ctx context.Context, repo *Repository) ([]*Repository, error) {
+	if !g.allBranches {
+		branch, _, err := g.client.GetRepoBranch(g.owner, repo.Repository, repo.Branch)
+		if err != nil {
+			return nil, err
+		}
+		return []*Repository{
+			{
+				Organization: repo.Organization,
+				Repository:   repo.Repository,
+				Branch:       repo.Branch,
+				URL:          repo.URL,
+				SHA:          branch.Commit.ID,
+				Labels:       repo.Labels,
+				RepositoryId: repo.RepositoryId,
+			},
+		}, nil
+	}
+	repos := []*Repository{}
+	opts := gitea.ListRepoBranchesOptions{}
+	branches, _, err := g.client.ListRepoBranches(g.owner, repo.Repository, opts)
+	if err != nil {
+		return nil, err
+	}
+	for _, branch := range branches {
+		repos = append(repos, &Repository{
+			Organization: repo.Organization,
+			Repository:   repo.Repository,
+			Branch:       branch.Name,
+			URL:          repo.URL,
+			SHA:          branch.Commit.ID,
+			Labels:       repo.Labels,
+			RepositoryId: repo.RepositoryId,
+		})
+	}
+	return repos, nil
+}
+
+func (g *GiteaProvider) ListRepos(ctx context.Context, cloneProtocol string) ([]*Repository, error) {
+	repos := []*Repository{}
+	repoOpts := gitea.ListOrgReposOptions{}
+	giteaRepos, _, err := g.client.ListOrgRepos(g.owner, repoOpts)
+	if err != nil {
+		return nil, err
+	}
+	for _, repo := range giteaRepos {
+		var url string
+		switch cloneProtocol {
+		// Default to SSH if unspecified (i.e. if "").
+		case "", "ssh":
+			url = repo.SSHURL
+		case "https":
+			url = repo.HTMLURL
+		default:
+			return nil, fmt.Errorf("unknown clone protocol for GitHub %v", cloneProtocol)
+		}
+		labelOpts := gitea.ListLabelsOptions{}
+		giteaLabels, _, err := g.client.ListRepoLabels(g.owner, repo.Name, labelOpts)
+		if err != nil {
+			return nil, err
+		}
+		labels := []string{}
+		for _, label := range giteaLabels {
+			labels = append(labels, label.Name)
+		}
+		repos = append(repos, &Repository{
+			Organization: g.owner,
+			Repository:   repo.Name,
+			Branch:       repo.DefaultBranch,
+			URL:          url,
+			Labels:       labels,
+			RepositoryId: int(repo.ID),
+		})
+	}
+	return repos, nil
+}
+
+func (g *GiteaProvider) RepoHasPath(ctx context.Context, repo *Repository, path string) (bool, error) {
+	_, resp, err := g.client.GetContents(repo.Organization, repo.Repository, repo.Branch, path)
+	if resp != nil && resp.StatusCode == 404 {
+		return false, nil
+	}
+	if fmt.Sprint(err) == "expect file, got directory" {
```

Comment:
also m not sure this is correct way to pass the test, does gitea have method to list both files and directory separately, plz throw some light!

Reply:
Oh I wish! the sdk was limited, this is the best I could come up with. I'll talk to the maintainers of the gitea-sdk, but it could be awhile before significant changes are made to account for directories.